src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java

```diff
@@ -0,0 +1,68 @@
+/*
+ *   Copyright (c) 2020-2021 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package chip.devicecontroller;
+
+@SuppressWarnings("serial")
+public class ConnectionFailureException extends ChipDeviceControllerException {
+  enum ConnectionState {
+    UNINITIALIZED, // Error state: OperationalSessionSetup is useless
+    NEEDS_ADDRESS, // No address known, lookup not started yet.
+    RESOLVING_ADDRESS, // Address lookup in progress.
+    HAS_ADDRESS, // Have an address, CASE handshake not started yet.
+    CONNECTING, // CASE handshake in progress.
+    SECURE_CONNECTED, // CASE session established.
+    WAITING_FOR_RETRY, // No address known, but a retry is pending.
+  }
+
+  private ConnectionState connectionState;
+
+  public ConnectionFailureException() {
+    super();
+  }
+
+  public ConnectionFailureException(long errorCode, int connectionState, String message) {
+    super(errorCode, message);
+    this.connectionState = mapIntToConnectionState(connectionState);
+  }
+
+  public ConnectionState getConnectionState() {
+    return connectionState;
+  }
+
+  // Helper method to map an int to ConnectionState enum
+  private ConnectionState mapIntToConnectionState(int value) {
```

Comment:
I think this would be easier to maintain if it's kept in the enum, see e.g. https://github.com/project-chip/connectedhomeip/blob/15f54da67d08f40b18d366433c492b7c1c8dd89a/src/controller/java/src/chip/onboardingpayload/CommissioningFlow.kt#L24

Reply:
Yes, that would be an ideal solution. However, we do face limitations because this class is generated from the native side, which only supports primary types as input parameters.

Reply:
Apologies I think you may have missed what I was saying - yes, from JNI, you'll need mapIntToConneciton state, but you'll want instead something like this:

```
 public enum ConnectionState {
    private final int number;
    public int getNumber() { return number }

    private ConnectionState(int number) { this.number = number }
    UNINITIALIZED(0 // Error state: OperationalSessionSetup is useless
    NEEDS_ADDRESS(1), // No address known, lookup not started yet.
...
  }
```

Then, here, you'd simply do:

```
for (state in ConnectionState.values()) {
   if (state.getNumber == value) {
      return state;
  }
}
```

(You can optimize with a map if necessary, but as this is a small list with a constrained size, for loop should do the trick)